### ISSUE TITLE
TypeError: object of type 'NoneType' has no len() & NotADirectoryError:

### ISSUE THREAD

**Issue by falahfakhri-Iraq** [johntruckenbrodt/pyroSAR#137]
- which operating system are you using?  
Windows 10, 
- which environment is pyroSAR running in?  
python 3.6, spyder 4.1.3 
- which version of pyroSAR are you using?  
0.12.1
one installed via pip or a clone of the GitHub repository?
pip install
-  which function of pyroSAR did you call with which parameters?  

I used your example to create the *.xml file it works fine, I only changed the GRD item, and the path

```Python
from pyroSAR.snap.auxil import parse_recipe, parse_node

workflow = parse_recipe('blank')

read = parse_node('Read')
read.parameters['file'] = 'D:\pyrosar_test\S1A_IW_GRDH_----------------------------.zip' # I think not necessary which item is.

read.parameters['formatName'] = 'SENTINEL-1'
workflow.insert_node(read)

tnr = parse_node('ThermalNoiseRemoval')
workflow.insert_node(tnr, before=read.id)

bnr = parse_node('Remove-GRD-Border-Noise')
bnr.parameters['selectedPolarisations'] = ['VV'] 

workflow.insert_node(bnr, before=tnr.id)

write = parse_node('Write')
write.parameters['file'] = 'outname'
write.parameters['formatName'] = 'BEAM-DIMAP'
workflow.insert_node(write, before=bnr.id)

workflow.write('D:\pyrosar_test\outname_proc')
from pyroSAR.snap.auxil import gpt
gpt('D:\\pyrosar_test\\outname_proc.xml', 'D:\\pyrosar_test')
```
AND I tried up also, 
```Python
gpt('D:\pyrosar_test\outname_proc.xml', 'D:\pyrosar_test')
```
- if applicable, which version of SNAP or GAMMA are you using in pyroSAR?
SNAP 8
- the full error message
```
Traceback (most recent call last):

  File "C:\Users\FALAH FAKHRI\untitled0.py", line 36, in <module>
    gpt('D:\\pyrosar_test\\outname_proc.xml', 'D:\\pyrosar_test')

  File "C:\Users\FALAH FAKHRI\AppData\Local\Programs\Python\Python36\lib\site-packages\pyroSAR\snap\auxil.py", line 343, in gpt
    while i < len(groups) - 1:

TypeError: object of type 'NoneType' has no len()
```

Then I used the *.xml which has been created in SNAP, it runs and I got the results I could open them, but gives this error, 
```Python
from pyroSAR.snap.auxil import gpt

gpt('D:\\pyrosar_test\\spdivpy201524.xml', 'D:\\pyrosar_test')
```
```
executing node sequence..
ThermalNoiseRemoval -> Apply-Orbit-File -> Calibration -> Speckle-Filter -> Speckle-Filter(2) -> Speckle-Divergence -> Speckle-Divergence(2)
Traceback (most recent call last):

  File "C:\Users\FALAH FAKHRI\untitled0.py", line 36, in <module>
    gpt('D:\\pyrosar_test\\spdivpy201524.xml', 'D:\\pyrosar_test')

  File "C:\Users\FALAH FAKHRI\AppData\Local\Programs\Python\Python36\lib\site-packages\pyroSAR\snap\auxil.py", line 420, in gpt
    shutil.rmtree(tmpname, onerror=windows_fileprefix)

  File "C:\Users\FALAH FAKHRI\AppData\Local\Programs\Python\Python36\lib\shutil.py", line 488, in rmtree
    return _rmtree_unsafe(path, onerror)

  File "C:\Users\FALAH FAKHRI\AppData\Local\Programs\Python\Python36\lib\shutil.py", line 370, in _rmtree_unsafe
    onerror(os.listdir, path, sys.exc_info())

  File "C:\Users\FALAH FAKHRI\AppData\Local\Programs\Python\Python36\lib\site-packages\pyroSAR\ancillary.py", line 303, in windows_fileprefix
    func(u'\\\\?\\' + path)

NotADirectoryError: [WinError 267] The directory name is invalid: '\\\\?\\D:\\pyrosar_test\\VH\\S1A_IW_GRDH_1SDV_20150824_VH_Orb_Cal_Spk_SpkDiv.dim'
```

**Comment by johntruckenbrodt** [comment]
Hi @falahfakhri-Iraq, thanks for reaching out. For a start I edited your post to highlight code and thus make it more readable.  
I'll soon look into it.

**Comment by falahfakhri-Iraq** [comment]
Dear @johntruckenbrodt, I installed anaconda, python v. 3.8.8, I installed all the dependencies and packages, however, I got similar error, by applying this code, and the *.xml file is yours from here, https://pyrosar.readthedocs.io/en/latest/general/snap.html
I only edit the tile name and path,  By the way, I used one backslash, and double backslash, 

`from pyroSAR.snap.auxil import gpt

gpt('D:\pyrosar_test\outname_proc.xml', 'D:\pyrosar_test')`  

and the error is, 

runfile('C:/Users/FALAH FAKHRI/untitled0.py', wdir='C:/Users/FALAH FAKHRI')
Traceback (most recent call last):

  File "C:\Users\FALAH FAKHRI\untitled0.py", line 10, in <module>
    gpt('D:\pyrosar_test\outname_proc.xml')

TypeError: gpt() missing 1 required positional argument: 'outdir'


runfile('C:/Users/FALAH FAKHRI/untitled0.py', wdir='C:/Users/FALAH FAKHRI')
Traceback (most recent call last):

  File "C:\Users\FALAH FAKHRI\untitled0.py", line 10, in <module>
    gpt('D:\pyrosar_test\outname_proc.xml', 'D:\pyrosar_test')

  File "C:\anaconda\envs\Deep Learning\lib\site-packages\pyroSAR\snap\auxil.py", line 343, in gpt
    while i < len(groups) - 1:

TypeError: object of type 'NoneType' has no len()


runfile('C:/Users/FALAH FAKHRI/untitled0.py', wdir='C:/Users/FALAH FAKHRI')
Traceback (most recent call last):

  File "C:\Users\FALAH FAKHRI\untitled0.py", line 10, in <module>
    gpt('D:\\pyrosar_test\\outname_proc.xml', 'D:\\pyrosar_test')

  File "C:\anaconda\envs\Deep Learning\lib\site-packages\pyroSAR\snap\auxil.py", line 343, in gpt
    while i < len(groups) - 1:

TypeError: object of type 'NoneType' has no len()

The following I used the *xml, has been created in SNAP

`from pyroSAR.snap.auxil import gpt

gpt('D:\pyrosar_test\spdivpy201524.xml', 'D:\pyrosar_test')`

As you see, it runs as below, 

runfile('C:/Users/FALAH FAKHRI/untitled0.py', wdir='C:/Users/FALAH FAKHRI')
executing node sequence..
ThermalNoiseRemoval -> Apply-Orbit-File -> Calibration -> Speckle-Filter -> Speckle-Filter(2) -> Speckle-Divergence -> Speckle-Divergence(2)

and then I got the following error, 


  File "C:\anaconda\envs\Deep Learning\lib\shutil.py", line 596, in _rmtree_unsafe
    with os.scandir(path) as scandir_it:

NotADirectoryError: [WinError 267] The directory name is invalid: 'D:\\pyrosar_test\\VH\\S1A_IW_GRDH_1SDV_20150824_VH_Orb_Cal_Spk_SpkDiv.dim'


During handling of the above exception, another exception occurred:

Traceback (most recent call last):

  File "C:\Users\FALAH FAKHRI\untitled0.py", line 10, in <module>
    gpt('D:\pyrosar_test\spdivpy201524.xml', 'D:\pyrosar_test')

  File "C:\anaconda\envs\Deep Learning\lib\site-packages\pyroSAR\snap\auxil.py", line 420, in gpt
    shutil.rmtree(tmpname, onerror=windows_fileprefix)

  File "C:\anaconda\envs\Deep Learning\lib\shutil.py", line 740, in rmtree
    return _rmtree_unsafe(path, onerror)

  File "C:\anaconda\envs\Deep Learning\lib\shutil.py", line 599, in _rmtree_unsafe
    onerror(os.scandir, path, sys.exc_info())

  File "C:\anaconda\envs\Deep Learning\lib\site-packages\pyroSAR\ancillary.py", line 303, in windows_fileprefix
    func(u'\\\\?\\' + path)

NotADirectoryError: [WinError 267] The directory name is invalid: '\\\\?\\D:\\pyrosar_test\\VH\\S1A_IW_GRDH_1SDV_20150824_VH_Orb_Cal_Spk_SpkDiv.dim'

**Comment by johntruckenbrodt** [comment]
Hi @falahfakhri-Iraq. Okay it seems there are multiple issues with the `gpt` function. It was designed to work well with the `geocode` function and it is not universal enough to use it in a different context. It expects the XML workflows to write files in a certain way. This error is definitely an issue in the `gpt` function:
```
TypeError: object of type 'NoneType' has no len()
```
Until this is fixed, I suggest you use the function `execute`, which plainly executes the workflow:
```Python
from pyroSAR.snap.auxil import execute
execute('workflow.xml')
```

Also this error is related to the incomplete design of the `gpt` function:
```
NotADirectoryError: [WinError 267] The directory name is invalid:
```
The way `gpt` works right now is to

- process all files to a temporary directory
- copy and rename all output files in ENVI format to `outdir`
- delete the temporary directory 

So basically `gpt` reads all file locations in the XML workflow as temporary locations and wants to delete them after having copied all relevant files. This does not work in your case and gives you an error.

I admit, it must be obvious from the documentation that `gpt` is limited and cannot easily be used for all tasks. Working on it.

The part with the missing argument is likely due to using single backslashes as directory names:
```
TypeError: gpt() missing 1 required positional argument: 'outdir'
```

**Comment by falahfakhri-Iraq** [comment]
Hi @johntruckenbrodt  
Thanks a lot to your efforts, meanwhile, I created the function which executes any *xml file has been created within SNAP, under python platform, 

The function link: https://github.com/falahfakhri-Iraq/gpt_launcher/blob/main/gpt_launcher.py
I think it's somehow similar to the execute method.

**Comment by johntruckenbrodt** [comment]
Hi @falahfakhri-Iraq. Great you have found a workaround in the meantime! Yes, this is very similar to `execute`.
I'll post updates to the bug fixing in the next days..

**Comment by johntruckenbrodt** [comment]
The following works now:
```python
import os
from pyroSAR.snap.auxil import parse_recipe, parse_node, gpt

workflow = parse_recipe('blank')

src = 'D:\\pyrosar_test\\S1A_IW_GRDH_----------------------------.zip'

outdir = 'D:\\pyrosar_test'
tmpdir = os.path.join(outdir, 'tmp')
outname = os.path.join(outdir, 'outname')
xmlfile = os.path.join(outdir, 'outname_proc.xml')

os.makedirs(outdir, exist_ok=True)

read = parse_node('Read')
read.parameters['file'] = src

read.parameters['formatName'] = 'SENTINEL-1'
workflow.insert_node(read)

tnr = parse_node('ThermalNoiseRemoval')
workflow.insert_node(tnr, before=read.id)

bnr = parse_node('Remove-GRD-Border-Noise')
bnr.parameters['selectedPolarisations'] = ['VV']

workflow.insert_node(bnr, before=tnr.id)

write = parse_node('Write')
write.parameters['file'] = outname
write.parameters['formatName'] = 'BEAM-DIMAP'
workflow.insert_node(write, before=bnr.id)

workflow.write(xmlfile)

gpt(xmlfile=xmlfile, cleanup=True, tmpdir=tmpdir)

```
